pallets/dapp-staking/src/lib.rs

```diff
@@ -1559,16 +1559,28 @@ pub mod pallet {
 
                         let preserved_bonus_status = staking_info.bonus_status;
                         // This need to be built before 'unstake', otherwise voting amount is converted into B&E amount
-                        let unstake_amount_iter: Vec<StakeAmount> =
+                        let unstake_amount =
                             Vec::from([staking_info.previous_staked, staking_info.staked])
                                 .into_iter()
                                 .filter(|stake_amount| !stake_amount.is_empty())
-                                .collect();
-                        let amount = staking_info.staked.total();
-
-                        staking_info.unstake(amount, current_era, protocol_state.subperiod());
+                                .map(|mut stake_amount| {
+                                    stake_amount.era = stake_amount.era.max(current_era);
+                                    stake_amount
+                                })
+                                .max_by(|a, b| a.total().cmp(&b.total()))
+                                // At least one value exists, otherwise we wouldn't be here.
+                                .ok_or(Error::<T>::InternalUnstakeError)?;
+
+                        let (unstake_amount_iter, _) = staking_info.unstake(
```

Comment:
Oh, so we weren't using the correct one 🙈

Reply:
Yes, I created a second one to preserve both stakes (voting + b&e) because the `unstake` method merge them when bonus is forfeited. But now, with a small reorg of the code, we use the correct and only one 🙂